integr8ly/integreatly-operator
# INTLY-3296 - use safe ownerReferences + finalizers on namespaces

JIRA: https://issues.jboss.org/browse/INTLY-3296

Motivation: product namespaces were removed on a kube-controller restart because of incorrect use of owner references. Installation CR, which is namespaced, owned non-namespaced resources (product ns, clusterRole and clusterRoleBinding for the backups, etc.) as well as namespaced resources in other namespaces (e.g. OperatorSourceConfig in openshift-marketplace ns). These ownerships are against Kubernetes garbage collection rules.

Solution: 
- all product namespaces will be cleaned up via a finalizer
- product namespace ownership is marked via "integreatly.org/installation-uid" label
- cluster resources related to backups(ClusterRole and ClusterRoleBinding) are owned by product namespace
- CatalogSourceConfigs is owned by product namespace
- backups now create a ClusterRole per product, so they can be removed independently

Verification:
- Install Integratly workshop type using this PR
- Check a CR - there should be a finilizer per each product with it's own namespace
- Check that CatalogSourceConfigs objects are own by respective namespaces (`oc get CatalogSourceConfigs -o yaml -n openshift-marketplace`)
- Check that backup cluster resources are owned by respecive namespaces (e.g. Codeready - `oc get ClusterRole codeready-backupjob -o yaml` + `oc get ClusterRoleBinding codeready-backupjob -o yaml`)
- Remove Installation CR and wait for a few minutes (in the mean time you can observe the CR, finilizers should be getting removed)
- Eventually all product namespaces and all resources they owned (e.g. CatalogSourceConfigs in openshift-marketplace ns) should be remove.
- Installation CR should be deleted as well


Note: I did couple of edits to unify code style in the places I was editing, and where it wouldn't change to many lines. Also I tweaked some seemingly unrelated code because unit tests were failing. Feel free to drop questions on individual pieces of code.
